### ISSUE TITLE
Panic: interface conversion: value.Named is nil, not *ir.BasicBlock

### ISSUE THREAD

**Issue by pwaller** [mewmew/l#4]
Using `(github.com/mewmew/l/asm).Parse` to parse the following results in a nil pointer dereference through a type assertion:

#### LLIR input to `Parse()`:

```llvm
define void @foo() {
	br label %l
l:
	ret void
}
```

#### Stack trace

```
panic: interface conversion: value.Named is nil, not *ir.BasicBlock

goroutine 1 [running]:
github.com/mewmew/l/asm/internal/irx.Translate.func7(0x121a920, 0xc003c7ff60)
	/Users/pwaller/go/src/github.com/mewmew/l/asm/internal/irx/module.go:270 +0xf4
github.com/mewmew/l/ir/irutil.(*walker).walkBeforeAfter(0xc003d17a38, 0x121a920, 0xc003c7ff60, 0x129bf28, 0xc003d17ce8)
	/Users/pwaller/go/src/github.com/mewmew/l/ir/irutil/walk.go:1530 +0x213
github.com/mewmew/l/ir/irutil.(*walker).walkBeforeAfter(0xc003d17a38, 0x12379a0, 0xc003c7ff60, 0x129bf28, 0xc003d17ce8)
	/Users/pwaller/go/src/github.com/mewmew/l/ir/irutil/walk.go:1126 +0x13baa
github.com/mewmew/l/ir/irutil.(*walker).walkBeforeAfter(0xc003d17a38, 0x1220020, 0xc0007605e8, 0x129bf28, 0xc003d17ce8)
	/Users/pwaller/go/src/github.com/mewmew/l/ir/irutil/walk.go:117 +0xb47d
github.com/mewmew/l/ir/irutil.(*walker).walkBeforeAfter(0xc003d17a38, 0x1258760, 0xc0007605c0, 0x129bf28, 0xc003d17ce8)
	/Users/pwaller/go/src/github.com/mewmew/l/ir/irutil/walk.go:894 +0x14f44
github.com/mewmew/l/ir/irutil.(*walker).walkBeforeAfter(0xc003d17a38, 0x121a920, 0xc000051280, 0x129bf28, 0xc003d17ce8)
	/Users/pwaller/go/src/github.com/mewmew/l/ir/irutil/walk.go:295 +0xe122
github.com/mewmew/l/ir/irutil.(*walker).walkBeforeAfter(0xc003d17a38, 0x1224200, 0xc003d04160, 0x129bf28, 0xc003d17ce8)
	/Users/pwaller/go/src/github.com/mewmew/l/ir/irutil/walk.go:1426 +0x2ec3
github.com/mewmew/l/ir/irutil.(*walker).walkBeforeAfter(0xc003d17a38, 0x121e6e0, 0xc0000b40c8, 0x129bf28, 0xc003d17ce8)
	/Users/pwaller/go/src/github.com/mewmew/l/ir/irutil/walk.go:539 +0x3223
github.com/mewmew/l/ir/irutil.WalkFuncBeforeAfter(0xc0000b4000, 0x129bf28, 0xc003d17ce8)
	/Users/pwaller/go/src/github.com/mewmew/l/ir/irutil/walk.go:63 +0x147
github.com/mewmew/l/ir/irutil.WalkFunc(0xc0000b4000, 0xc003d17ce8)
	/Users/pwaller/go/src/github.com/mewmew/l/ir/irutil/walk.go:29 +0x41
github.com/mewmew/l/asm/internal/irx.Translate(0xc003c7ffe0, 0x30, 0x600, 0xc003c7ffe0)
	/Users/pwaller/go/src/github.com/mewmew/l/asm/internal/irx/module.go:274 +0xabb
github.com/mewmew/l/asm.ParseBytes(0xc003d00000, 0x30, 0x600, 0x30, 0x600, 0x0)
	/Users/pwaller/go/src/github.com/mewmew/l/asm/asm.go:43 +0x60
github.com/mewmew/l/asm.Parse(0x12c0780, 0xc00000c0b8, 0xc00000c0b8, 0x0, 0x0)
	/Users/pwaller/go/src/github.com/mewmew/l/asm/asm.go:32 +0x9f
```

#### Failing code:

https://github.com/mewmew/l/blob/935e562f92c3ad2bd406e7c262c0497afaddcce8/asm/internal/irx/module.go#L270

**Comment by mewmew** [comment]
Good catch @pwaller. Will have to take a look at this. I just got back to Sweden, and will probably get down to coding in a few weeks. Ping me if I have not had a chance to look at it before then.

Happy coding!
Robin

**Comment by pwaller** [comment]
I've made a commit in my fork (above) which makes the error go away - I can't speak to its correctness, so I haven't submitted it as a PR yet. FYI I'm away for a bit, but I should be communicating again within a couple of weeks.

**Comment by mewmew** [comment]
> I've made a commit in my fork (above) which makes the error go away - I can't speak to its correctness, so I haven't submitted it as a PR yet. FYI I'm away for a bit, but I should be communicating again within a couple of weeks.

Thanks! I'll take a look. Have a good trip and talk with you in a couple of weeks :)

**Comment by pwaller** [comment]
Hi @mewmew - can we get this one in? Also, did you get my email sent on 4th October around 0930 UTC?

**Comment by mewmew** [comment]
Thanks a lot for the minimal test case! I think I've tracked down the original issue.

In short, given:

```llvm
define void @foo() {
	br label %foo
foo:
	ret void
}
```

The name of the second basic block was stored as `oo:` rather than `foo` (notice the trailing colon). So basically, this must steam from an incorrect slice operation (e.g. `name[len(prefix):]` rather than `name[:len(name)-len(suffix)]`.

```
block: label %"oo\3A"
```

The fix should be easy, and is upcoming. Thanks for the patience :)

**Comment by mewmew** [comment]
Fix submitted in c756be7.

@pwaller would you please try your original code and see if it works? The minimal test case you posted works now.

```diff
diff --git a/asm/internal/astx/identifier.go b/asm/internal/astx/identifier.go
index f4050c4..1ae66a1 100644
--- a/asm/internal/astx/identifier.go
+++ b/asm/internal/astx/identifier.go
@@ -50,7 +50,7 @@ func NewLabelIdent(tok interface{}) (*ast.LabelIdent, error) {
        if !strings.HasSuffix(ident, suffix) {
                return nil, errors.Errorf("invalid label identifier %q; missing '%v' suffix", ident, suffix)
        }
-       ident = ident[len(suffix):]
+       ident = ident[:len(ident)-len(suffix)]
        if strings.HasPrefix(ident, `"`) && strings.HasSuffix(ident, `"`) {
                ident = enc.Unquote(ident)
        }
```
